### PR TITLE
bug 1184662 - Add support for Safari

### PR DESCRIPTION
**Pull request by darkwing**
To Do:

- [x] Add instructions for testing with Safari
- [x] Remove the `sleep` calls required to make Safari auth tests work

### CODE REVIEW

File: tests/ui/tests/lib/capabilities.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        },
+
+        crossbrowserConfirm: function(remote) {
+            // Firefox and Chrome react more reliably to elements with the ENTER key pressed
+            // Safari, on the other hand, only responds to click
+
+            var gbn = this.getBrowserName(remote);
+
+            return function(element) {
+                if(gbn === 'safari') {
```
Comment (groovecoder) [inline tests/ui/tests/lib/capabilities.js]: Ouch. :cry:

File: tests/ui/tests/lib/login.js [hunk 2]
```diff
@@ -72,6 +72,11 @@ define([
                     .then(function(handles) {
 
                         return remote.switchToWindow(handles[1])
+
+                            // FOR SAFARI Added this for Safari -- it isn't finding the #authentication_email without it
+                            // Could be a page load issue, may need to add polling for the element first
+                            .sleep(5000)
```
Comment (groovecoder) [inline tests/ui/tests/lib/login.js]: Could we wrap this in something similar to `this.getBrowserName(remote) === 'safari'` ?

File: tests/ui/tests/lib/login.js [hunk 3]
```diff
@@ -168,6 +170,9 @@ define([
                                 });
                             })
 
+
+                            .sleep(4000) // FOR SAFARI
```
Comment (groovecoder) [inline tests/ui/tests/lib/login.js]: Again, could we only sleep here for Safari?

**Comment by groovecoder** [comment]
https://github.com/mozilla/kuma/pull/3531 is merged if you want to add the Safari instructions to the newly updated `tests-ui.rst`.

**Comment by groovecoder** [comment]
I added the Safari Driver and ran the tests but I got lots of errors and failures. Seems like my Safari wouldn't open the Persona window at all?

File: tests/ui/tests/lib/login.js (hunk 2 above)
Comment (darkwing) [inline tests/ui/tests/lib/login.js]: Done locally, hadn't pushed yet.

File: tests/ui/tests/lib/login.js (hunk 3 above)
Comment (darkwing) [inline tests/ui/tests/lib/login.js]: Done locally, hadn't pushed yet.

**Comment by darkwing** [comment]
@groovecoder You have to open Safari manually, go to the Preferences pane, and disable the Popup Blocker manually :(

File: tests/ui/tests/lib/capabilities.js (hunk 1 above)
Comment (darkwing) [inline tests/ui/tests/lib/capabilities.js]: Sucks but not as bad as you think -- we control the names of the browsers so it's fine.  We're writing for Selenium/node, not the browser.

**Comment by groovecoder** [comment]
Cool.
```
safari on any platform: 1/31 tests failed

× safari on any platform - dashboards - [requires-login][requires-destructive] Revision dashboard banning (30.01s)
CancelError: Timeout reached on safari on any platform - dashboards - [requires-login][requires-destructive] Revision dashboard banning
  at null._onTimeout  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
  at Command.then  <node_modules/intern/node_modules/leadfoot/Command.js:542:10>
  at Command  <node_modules/intern/node_modules/leadfoot/Command.js:566:15>
  at <node_modules/intern/lib/Test.js:236:37>
  at Promise.cancel  <node_modules/intern/node_modules/dojo/Promise.ts:343:37>
  at null._onTimeout  <node_modules/intern/lib/Test.js:164:21>
  at Timer.listOnTimeout  <timers.js:110:15>
```
Though the particular functionality seemed to work?

The last annoying thing is that Safari gives a certificate warning dialog when it first starts to run the test suite. So I have to manually click to continue before it will start running. :neutral_face:

File: tests/ui/tests/lib/capabilities.js (hunk 1 above)
Comment (groovecoder) [inline tests/ui/tests/lib/capabilities.js]: I mostly meant the fact that we have to introduce browser-specific branching. :cry:

**Comment by darkwing** [comment]
If you click the "Show Certificate" button, you can "Always trust" the site.

I've updated that test to use the Safari sleep shim because it pops up a new window :(.  Works now though.

**Comment by darkwing** [comment]
Instructions/docs added!  Ready for another look.

**Comment by groovecoder** [comment]
Running with:
```
./node_modules/.bin/intern-runner config=intern-local d=developer-local.allizom.org u=firefoxos@davidwalsh.name b=safari descructive=true wd=User:groovecoder p=
```
Got:
```
TOTAL: tested 1 platforms, 9/31 tests failed
```
But the details only list 3 with `×`:
```
× safari on any platform - header - The header search box expands and contracts correctly (30.022s)
× safari on any platform - dashboards - [requires-login][requires-destructive] Revision dashboard banning (0.874s)
× safari on any platform - wiki - [requires-login] Logging in for wiki tests (30.004s)
```

**Comment by darkwing** [comment]
I'll run another check but the first failure tells me you didn't manually make Safari the foremost window so that the animation runs, like Firefox :/

**Comment by darkwing** [comment]
Also, if you don't finish a test run, Safari will keep your last login to MDN active :(  Make sure you aren't logged in to begin with.

**Comment by groovecoder** [comment]
I signed out of MDN and Persona in Safari, but it got worse. :(
```
safari on any platform: 13/31 tests failed
```
And I killed it after so many `requires-login` tests kept hitting their 30s timeout.

**Comment by darkwing** [comment]
Just retested and everything worked for me.  Can you tell me:

- [ ] - What version of Safari WebDriver Extension?  I have 2.45.0
- [ ] - What version of Intern?  I upgraded to 3.0.5
- [ ] - When login fails, please click back to the Persona popup and let me know what you see.  Sometimes I see an empty window...and I don't know what I can do about that.

**Comment by groovecoder** [comment]
Changed tack to run this in BrowserStack with this in the `intern-browserstack.js`:
```
         { browserName: 'safari', version: '8' },
```
Got this:
```
Saf 8 MAC:  [✓×✓✓✓✓✓✓✓✓✓✓✓✓×××✓✓××××××✓✓×/  ] 28/31, 11 fail
```
Deets:
```
✓ safari 8 on any platform - footer - Changing the footer's language selector changes locale via URL
✓ safari 8 on any platform - auth - Hovering over the header nav widget opens submenu
✓ safari 8 on any platform - demos - The 3 featured "demos-main" element is present
✓ safari 8 on any platform - demos - The demo search form is present
✓ safari 8 on any platform - env - jQuery object is provided to MDN
✓ safari 8 on any platform - env - mdn object is provided to MDN
✓ safari 8 on any platform - env - waffle object is provided to MDN
✓ safari 8 on any platform - env - gettext object is provided to MDN
✓ safari 8 on any platform - header - Pressing [ENTER] submits the header search box
✓ safari 8 on any platform - home - Homepage search form displays and accepts text, [ENTER] key submits form
✓ safari 8 on any platform - home - Hacks posts display properly
✓ safari 8 on any platform - home - Large search field does not display on mobile and lower
✓ safari 8 on any platform - header - Hovering over Zones menu displays submenu
✓ safari 8 on any platform - auth - Clicking on the GitHub icon initializes GitHub login process
✓ safari 8 on any platform - auth - Sign in icons are hidden from header widget on smaller screens
✓ safari 8 on any platform - wiki - [requires-login][requires-destructive] Created page displays review requests
✓ safari 8 on any platform - dashboards - Revision dashboard renders
× safari 8 on any platform - auth - [requires-login] Logging in with Persona with real credentials works
CancelError: Timeout reached on safari 8 on any platform - auth - [requires-login] Logging in with Persona with real credentials works
  at null.<anonymous>  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
× safari 8 on any platform - dashboards - [requires-login][requires-destructive] Revision dashboard banning
CancelError: Timeout reached on safari 8 on any platform - dashboards - [requires-login][requires-destructive] Revision dashboard banning
  at null._onTimeout  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
  at Command.then  <node_modules/intern/node_modules/leadfoot/Command.js:542:10>
  at Command  <node_modules/intern/node_modules/leadfoot/Command.js:566:15>
  at <node_modules/intern/lib/Test.js:236:37>
  at Promise.cancel  <node_modules/intern/node_modules/dojo/Promise.ts:343:37>
  at null._onTimeout  <node_modules/intern/lib/Test.js:164:21>
  at Timer.listOnTimeout  <timers.js:110:15>
× safari 8 on any platform - header - The header search box expands and contracts correctly
CancelError: Timeout reached on safari 8 on any platform - header - The header search box expands and contracts correctly
  at null._onTimeout  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
  at Command.target.(anonymous function) [as executeAsync]  <node_modules/intern/node_modules/leadfoot/Command.js:610:11>
  at Test.registerSuite.The header search box expands and contracts correctly [as test]  </Users/lcrouch/code/kuma/tests/ui/tests/header.js:84:21>
  at <node_modules/intern/lib/Test.js:211:24>
  at <node_modules/intern/node_modules/dojo/Promise.ts:393:15>
  at runCallbacks  <node_modules/intern/node_modules/dojo/Promise.ts:11:11>
  at <node_modules/intern/node_modules/dojo/Promise.ts:317:4>
  at run  <node_modules/intern/node_modules/dojo/Promise.ts:237:7>
  at <node_modules/intern/node_modules/dojo/nextTick.ts:44:3>
  at process._tickCallback  <node.js:355:11>
  at Command.then  <node_modules/intern/node_modules/leadfoot/Command.js:542:10>
  at Command  <node_modules/intern/node_modules/leadfoot/Command.js:566:15>
  at <node_modules/intern/lib/Test.js:236:37>
  at Promise.cancel  <node_modules/intern/node_modules/dojo/Promise.ts:343:37>
  at null._onTimeout  <node_modules/intern/lib/Test.js:164:21>
  at Timer.listOnTimeout  <timers.js:110:15>
× safari 8 on any platform - wiki - [requires-login] Logging in for wiki tests
CancelError: Timeout reached on safari 8 on any platform - wiki - [requires-login] Logging in for wiki tests
  at null.<anonymous>  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
× safari 8 on any platform - wiki - [requires-login] The new document screen passes all the checks
CancelError: Timeout reached on safari 8 on any platform - wiki - [requires-login] The new document screen passes all the checks
  at null._onTimeout  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
  at Command.then  <node_modules/intern/node_modules/leadfoot/Command.js:542:10>
  at Command  <node_modules/intern/node_modules/leadfoot/Command.js:566:15>
  at <node_modules/intern/lib/Test.js:236:37>
  at Promise.cancel  <node_modules/intern/node_modules/dojo/Promise.ts:343:37>
  at null._onTimeout  <node_modules/intern/lib/Test.js:164:21>
  at Timer.listOnTimeout  <timers.js:110:15>
× safari 8 on any platform - wiki - [requires-login] IFRAME elements are not allowed by CKEditor
CancelError: Timeout reached on safari 8 on any platform - wiki - [requires-login] IFRAME elements are not allowed by CKEditor
  at null._onTimeout  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
  at Command.then  <node_modules/intern/node_modules/leadfoot/Command.js:542:10>
  at Command  <node_modules/intern/node_modules/leadfoot/Command.js:566:15>
  at <node_modules/intern/lib/Test.js:236:37>
  at Promise.cancel  <node_modules/intern/node_modules/dojo/Promise.ts:343:37>
  at null._onTimeout  <node_modules/intern/lib/Test.js:164:21>
  at Timer.listOnTimeout  <timers.js:110:15>
× safari 8 on any platform - wiki - [requires-login] The new document-template screen passes all the checks
CancelError: Timeout reached on safari 8 on any platform - wiki - [requires-login] The new document-template screen passes all the checks
  at null._onTimeout  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
  at Command.then  <node_modules/intern/node_modules/leadfoot/Command.js:542:10>
  at Command  <node_modules/intern/node_modules/leadfoot/Command.js:566:15>
  at <node_modules/intern/lib/Test.js:236:37>
  at Promise.cancel  <node_modules/intern/node_modules/dojo/Promise.ts:343:37>
  at null._onTimeout  <node_modules/intern/lib/Test.js:164:21>
  at Timer.listOnTimeout  <timers.js:110:15>
× safari 8 on any platform - wiki - [requires-login][requires-destructive] Can create a document, button becomes enabled as soon as updates made
CancelError: Timeout reached on safari 8 on any platform - wiki - [requires-login][requires-destructive] Can create a document, button becomes enabled as soon as updates made
  at null._onTimeout  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
  at Command.then  <node_modules/intern/node_modules/leadfoot/Command.js:542:10>
  at Command  <node_modules/intern/node_modules/leadfoot/Command.js:566:15>
  at <node_modules/intern/lib/Test.js:236:37>
  at Promise.cancel  <node_modules/intern/node_modules/dojo/Promise.ts:343:37>
  at null._onTimeout  <node_modules/intern/lib/Test.js:164:21>
  at Timer.listOnTimeout  <timers.js:110:15>
× safari 8 on any platform - wiki - [requires-login][requires-destructive] Created page passes TOC tests
NoSuchElement: [POST http://(redacted)@hub.browserstack.com:4444/wd/hub/session/1f0476bd66e612c24e3c8ae5713a50bd35b75786/element / {"using":"css selector","value":"#toc"}] Could not find element: {"using":"css selector","value":"#toc"} (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 8 milliseconds
For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html
Build info: version: '2.45.0', revision: '5017cb8', time: '2015-02-26 23:59:50'
System info: host: 'mac-208-78-105-29.browserstack.com', ip: '208.78.105.29', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.10.5', java.version: '1.8.0_51'
Driver info: org.openqa.selenium.safari.SafariDriver
Capabilities [{browserName=safari, takesScreenshot=true, javascriptEnabled=true, version=8.0.8, cssSelectorsEnabled=true, platform=MAC, secureSsl=true}]
Session ID: null
*** Element info: {Using=css selector, value=#toc}
  at ProxiedSession._post  <node_modules/intern/node_modules/leadfoot/Session.js:59:30>
  at ProxiedSession.Session.find  <node_modules/intern/node_modules/leadfoot/Session.js:1055:15>
  at Command.<anonymous>  <node_modules/intern/node_modules/leadfoot/Command.js:42:36>
  at <node_modules/intern/node_modules/dojo/Promise.ts:393:15>
  at run  <node_modules/intern/node_modules/dojo/Promise.ts:237:7>
  at <node_modules/intern/node_modules/dojo/nextTick.ts:44:3>
  at process._tickCallback  <node.js:355:11>
  at Command.find  <node_modules/intern/node_modules/leadfoot/Command.js:23:10>
  at Command.prototype.(anonymous function) [as findByCssSelector]  <node_modules/intern/node_modules/leadfoot/lib/strategies.js:24:16>
  at Test.registerSuite.[requires-login][requires-destructive] Created page passes TOC tests [as test]  </Users/lcrouch/code/kuma/tests/ui/tests/wiki.js:249:21>
  at <node_modules/intern/lib/Test.js:211:24>
  at <node_modules/intern/node_modules/dojo/Promise.ts:393:15>
  at runCallbacks  <node_modules/intern/node_modules/dojo/Promise.ts:11:11>
  at <node_modules/intern/node_modules/dojo/Promise.ts:317:4>
  at run  <node_modules/intern/node_modules/dojo/Promise.ts:237:7>
  at <node_modules/intern/node_modules/dojo/nextTick.ts:44:3>
  at process._tickCallback  <node.js:355:11>
× safari 8 on any platform - auth - Clicking Persona link opens new window
CancelError: Timeout reached on safari 8 on any platform - auth - Clicking Persona link opens new window
  at null._onTimeout  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
  at Command.then  <node_modules/intern/node_modules/leadfoot/Command.js:542:10>
  at Test.registerSuite.Clicking Persona link opens new window [as test]  </Users/lcrouch/code/kuma/tests/ui/tests/auth.js:47:21>
  at <node_modules/intern/lib/Test.js:211:24>
  at <node_modules/intern/node_modules/dojo/Promise.ts:393:15>
  at runCallbacks  <node_modules/intern/node_modules/dojo/Promise.ts:11:11>
  at <node_modules/intern/node_modules/dojo/Promise.ts:317:4>
  at run  <node_modules/intern/node_modules/dojo/Promise.ts:237:7>
  at <node_modules/intern/node_modules/dojo/nextTick.ts:44:3>
  at process._tickCallback  <node.js:355:11>
  at Command.then  <node_modules/intern/node_modules/leadfoot/Command.js:542:10>
  at Command  <node_modules/intern/node_modules/leadfoot/Command.js:566:15>
  at <node_modules/intern/lib/Test.js:236:37>
  at Promise.cancel  <node_modules/intern/node_modules/dojo/Promise.ts:343:37>
  at null._onTimeout  <node_modules/intern/lib/Test.js:164:21>
  at Timer.listOnTimeout  <timers.js:110:15>
× safari 8 on any platform - auth - Logging in with Persona for the first time sends user to registration page
CancelError: Timeout reached on safari 8 on any platform - auth - Logging in with Persona for the first time sends user to registration page
  at null.<anonymous>  <node_modules/intern/lib/Test.js:162:20>
  at Timer.listOnTimeout  <timers.js:110:15>
! [POST http://(redacted)@hub.browserstack.com:4444/wd/hub/session/1f0476bd66e612c24e3c8ae5713a50bd35b75786/execute / {"script":"return (function getCoverageData(coverageVariable) {\n\t\tvar coverageData = (function () { return this; })()[coverageVariable];\n\t\treturn coverageData && JSON.stringify(coverageData);\n\t}).apply(this, arguments);","args":["__internCoverage"]}] Timed out awaiting response to command "executeScript" after 30002 ms (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 30.01 seconds
Build info: version: '2.45.0', revision: '5017cb8', time: '2015-02-26 23:59:50'
System info: host: 'mac-208-78-105-29.browserstack.com', ip: '208.78.105.29', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.10.5', java.version: '1.8.0_51'
Driver info: org.openqa.selenium.safari.SafariDriver
Capabilities [{browserName=safari, takesScreenshot=true, javascriptEnabled=true, version=8.0.8, cssSelectorsEnabled=true, platform=MAC, secureSsl=true}]
Session ID: null
ScriptTimeout: [POST http://(redacted)@hub.browserstack.com:4444/wd/hub/session/1f0476bd66e612c24e3c8ae5713a50bd35b75786/execute / {"script":"return (function getCoverageData(coverageVariable) {\n\t\tvar coverageData = (function () { return this; })()[coverageVariable];\n\t\treturn coverageData && JSON.stringify(coverageData);\n\t}).apply(this, arguments);","args":["__internCoverage"]}] Timed out awaiting response to command "executeScript" after 30002 ms (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 30.01 seconds
Build info: version: '2.45.0', revision: '5017cb8', time: '2015-02-26 23:59:50'
System info: host: 'mac-208-78-105-29.browserstack.com', ip: '208.78.105.29', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.10.5', java.version: '1.8.0_51'
Driver info: org.openqa.selenium.safari.SafariDriver
Capabilities [{browserName=safari, takesScreenshot=true, javascriptEnabled=true, version=8.0.8, cssSelectorsEnabled=true, platform=MAC, secureSsl=true}]
Session ID: null
  at ProxiedSession._post  <node_modules/intern/node_modules/leadfoot/Session.js:59:30>
  at ProxiedSession.Session.execute  <node_modules/intern/node_modules/leadfoot/Session.js:583:15>
  at ProxiedSession.lang.mixin.get  <node_modules/intern/lib/ProxiedSession.js:81:17>
  at Command.<anonymous>  <node_modules/intern/node_modules/leadfoot/Command.js:635:19>
  at <node_modules/intern/node_modules/dojo/Promise.ts:393:15>
  at run  <node_modules/intern/node_modules/dojo/Promise.ts:237:7>
  at <node_modules/intern/node_modules/dojo/nextTick.ts:44:3>
  at process._tickCallback  <node.js:355:11>
  at Command.target.(anonymous function) [as get]  <node_modules/intern/node_modules/leadfoot/Command.js:610:11>
  at ClientSuite.lang.mixin.run  <node_modules/intern/lib/ClientSuite.js:149:6>
  at runWithCatch  <node_modules/intern/lib/Suite.js:314:21>
  at next  <node_modules/intern/lib/Suite.js:322:18>
  at <node_modules/intern/lib/Suite.js:345:6>
  at new Promise  <node_modules/intern/node_modules/dojo/Promise.ts:411:3>
  at runTests  <node_modules/intern/lib/Suite.js:278:12>
  at <node_modules/intern/node_modules/dojo/Promise.ts:393:15>
  at runCallbacks  <node_modules/intern/node_modules/dojo/Promise.ts:11:11>
  at <node_modules/intern/node_modules/dojo/Promise.ts:317:4>
```

**Comment by darkwing** [comment]
Those failures are all symptoms we can reproduce locally:  

1.  The header test fails because Safari is not the "front"/selected window
2.  The other tests (probably) fail due to popup blocker being enabled.  Going to investigate BrowserStack + Safari

**Comment by groovecoder** [comment]
FWIW, I also have Safari driver 2.45.0 and I just upgraded to Intern 3.0.5. I still got lots of the same failures, and my persona window seems to hang like this:
safari-persona-intern

**Comment by groovecoder** [comment]
Note: this is merge-able as a work-in-progress for Safari support. Until we iron out all the Safari issues, I'll simply just suggest that we don't block deployments on Safari by keeping Safari out of our BrowserStack config.

**Pull request merged**